lib/vfscore/main.c

```diff
@@ -1158,39 +1234,91 @@ struct dirent *readdir(DIR *dir)
 
 int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
 {
-	int error;
+	/**
+	 * `sys_readdir()` will end up calling `VOP_READDIR`, which expects
+	 * a dirent64 structure instead of the `struct dirent` entry parameter.
+	 */
+	struct dirent64 entry64;
 	struct vfscore_file *fp;
+	int error;
+
+	*result = NULL;
 
-	trace_vfs_readdir(dir->fd, entry);
-	error = fget(dir->fd, &fp);
-	if (!error) {
-		error = sys_readdir(fp, entry);
-		fdrop(fp);
-		if (error) {
-			trace_vfs_readdir_err(error);
-		} else {
-			trace_vfs_readdir_ret();
-		}
-	}
 	// Our dirent has (like Linux) a d_reclen field, but a constant size.
 	entry->d_reclen = sizeof(*entry);
 
-	if (error) {
-		*result = NULL;
-	} else {
-		*result = entry;
+	trace_vfs_readdir(dir->fd, entry);
+	error = fget(dir->fd, &fp);
+	if (unlikely(error))
+		return error;
+
+	error = sys_readdir(fp, &entry64);
+	fdrop(fp);
+	if (unlikely(error)) {
+		trace_vfs_readdir_err(error);
+
+		return error == ENOENT ? 0 : error;
 	}
-	return error == ENOENT ? 0 : error;
+
+	trace_vfs_readdir_ret();
+
+	entry->d_ino = entry64.d_ino;
+	entry->d_type = entry64.d_type;
+	entry->d_off = entry64.d_off;
+	memcpy(entry->d_name, entry64.d_name, sizeof(entry64.d_name));
+
+	*result = entry;
+
+	return 0;
 }
+#else
+struct dirent *readdir(DIR *dir) __attribute__((alias("readdir64")));
+int readdir_r(DIR *dir, struct dirent *entry, struct dirent **result)
+	 __attribute__((alias("readdir64_r")));
+#endif /* CONFIG_LIBVFSCORE_NONLARGEFILE */
 
-// FIXME: in 64bit dirent64 and dirent are identical, so it's safe to alias
 #undef readdir64_r
```

Comment:
Do we need to guard this `#undef` with `#ifdef`?

Reply:
Yes, it was already there so I missed it.